src/osd/OSD.cc

```diff
@@ -365,83 +366,86 @@ void OSDService::identify_splits_and_merges(
     queue.pop_front();
     did.insert(cur);
     unsigned pgnum = old_pgnum;
-    for (auto q = p->second.lower_bound(old_map->get_epoch());
-	 q != p->second.end() &&
-	   q->first <= new_map->get_epoch();
-	 ++q) {
-      if (pgnum < q->second) {
+    for (auto map_iter = pool_pg_num_history_map.lower_bound(old_map->get_epoch());
+	 map_iter != pool_pg_num_history_map.end();
+	 ++map_iter) {
+      auto& [new_epoch, new_pgnum] = *map_iter;
+      if (new_epoch <= new_map->get_epoch()) {
```

Comment:
The for loop definition above causes map_iter to take on all elements with keys in the set [old_map->get_epoch(), infinity).  This condition here skips every map in the  set [0, new_map->get_epoch].  This leaves (new_map->get_epoch(), infinity).  Why not modify the for loop start bound to pool_pg_num_history_map.upper_bound(new_map->get_epoch())?

Reply:
In a different commit, though, since it's a behavior difference.

Reply:
I got the condition wrong, fixed. It's there to avoid handling any information recorded beyond  `new_map->get_epoch()` (added comment).